In [49]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import test
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 8: Treetop Tree House ---</h2><p>The expedition comes across a peculiar patch of tall trees all planted carefully in a grid. The Elves explain that a previous expedition planted these trees as a reforestation effort. Now, they're curious if this would be a good location for a <a href="https://en.wikipedia.org/wiki/Tree_house" target="_blank">tree house</a>.</p>
<p>First, determine whether there is enough tree cover here to keep a tree house <em>hidden</em>. To do this, you need to count the number of trees that are <em>visible from outside the grid</em> when looking directly along a row or column.</p>
<p>The Elves have already launched a <a href="https://en.wikipedia.org/wiki/Quadcopter" target="_blank">quadcopter</a> to generate a map with the height of each tree (<span title="The Elves have already launched a quadcopter (your puzzle input).">your puzzle input</span>). For example:</p>
<pre><code>30373
25512
65332
33549
35390
</code></pre>
<p>Each tree is represented as a single digit whose value is its height, where <code>0</code> is the shortest and <code>9</code> is the tallest.</p>
<p>A tree is <em>visible</em> if all of the other trees between it and an edge of the grid are <em>shorter</em> than it. Only consider trees in the same row or column; that is, only look up, down, left, or right from any given tree.</p>
<p>All of the trees around the edge of the grid are <em>visible</em> - since they are already on the edge, there are no trees to block the view. In this example, that only leaves the <em>interior nine trees</em> to consider:</p>
<ul>
<li>The top-left <code>5</code> is <em>visible</em> from the left and top. (It isn't visible from the right or bottom since other trees of height <code>5</code> are in the way.)</li>
<li>The top-middle <code>5</code> is <em>visible</em> from the top and right.</li>
<li>The top-right <code>1</code> is not visible from any direction; for it to be visible, there would need to only be trees of height <em>0</em> between it and an edge.</li>
<li>The left-middle <code>5</code> is <em>visible</em>, but only from the right.</li>
<li>The center <code>3</code> is not visible from any direction; for it to be visible, there would need to be only trees of at most height <code>2</code> between it and an edge.</li>
<li>The right-middle <code>3</code> is <em>visible</em> from the right.</li>
<li>In the bottom row, the middle <code>5</code> is <em>visible</em>, but the <code>3</code> and <code>4</code> are not.</li>
</ul>
<p>With 16 trees visible on the edge and another 5 visible in the interior, a total of <code><em>21</em></code> trees are visible in this arrangement.</p>
<p>Consider your map; <em>how many trees are visible from outside the grid?</em></p>
</article>


In [50]:
class Grid:
    def __init__(self, s: str) -> None:
        self.grid = [[int(c) for c in l.strip()] for l in s.strip().splitlines()]
        rows, cols = len(self.grid), len(self.grid[0])
        self.ids = [[r * cols + c for c in range(cols)] for r in range(rows)]

    def rotate_90_degrees(self) -> None:
        self.grid = list(zip(*self.grid[::-1]))
        self.ids = list(zip(*self.ids[::-1]))

    def visible(self) -> int:
        # visinle from left
        visible = self.collect_visible_from_left_to_right()
        # visinle from bottom
        self.rotate_90_degrees()
        visible |= self.collect_visible_from_left_to_right()
        # visinle from right
        self.rotate_90_degrees()
        visible |= self.collect_visible_from_left_to_right()
        # visinle from top
        self.rotate_90_degrees()
        visible |= self.collect_visible_from_left_to_right()

        return len(visible)

    def collect_visible_from_left_to_right(self) -> set[int]:
        visible = set()
        for row_gr, row_id in zip(self.grid, self.ids):
            max_length = -1
            for tree, identifier in zip(row_gr, row_id):
                if tree > max_length:
                    max_length = tree
                    visible.add(identifier)
        return visible

    def __str__(self) -> str:
        return "\n".join(f"{''.join(str(i) for i in t)}" for t in self.grid)


tests = [
    {
        "name": "Example",
        "s": """
            30373
            25512
            65332
            33549
            35390
        """,
        "expected": 21,
    },
]


@test(tests=tests)
def test_part_I(s: str) -> int:
    return Grid(s).visible()


Test Example passed, for test_part_I.
Success


In [51]:
with open("../input/day8.txt") as f:
    puzzle = f.read()

print(f"Part 1: {Grid(puzzle).visible()}")

Part 1: 1849


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>1849</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>


<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Content with the amount of tree cover available, the Elves just need to know the best spot to build their tree house: they would like to be able to see a lot of <em>trees</em>.</p>
<p>To measure the viewing distance from a given tree, look up, down, left, and right from that tree; stop if you reach an edge or at the first tree that is the same height or taller than the tree under consideration. (If a tree is right on the edge, at least one of its viewing distances will be zero.)</p>
<p>The Elves don't care about distant trees taller than those found by the rules above; the proposed tree house has large <a href="https://en.wikipedia.org/wiki/Eaves" target="_blank">eaves</a> to keep it dry, so they wouldn't be able to see higher than the tree house anyway.</p>
<p>In the example above, consider the middle <code>5</code> in the second row:</p>
<pre><code>30373
25<em>5</em>12
65332
33549
35390
</code></pre>
<ul>
<li>Looking up, its view is not blocked; it can see <code><em>1</em></code> tree (of height <code>3</code>).</li>
<li>Looking left, its view is blocked immediately; it can see only <code><em>1</em></code> tree (of height <code>5</code>, right next to it).</li>
<li>Looking right, its view is not blocked; it can see <code><em>2</em></code> trees.</li>
<li>Looking down, its view is blocked eventually; it can see <code><em>2</em></code> trees (one of height <code>3</code>, then the tree of height <code>5</code> that blocks its view).</li>
</ul>
<p>A tree's <em>scenic score</em> is found by <em>multiplying together</em> its viewing distance in each of the four directions. For this tree, this is <code><em>4</em></code> (found by multiplying <code>1 * 1 * 2 * 2</code>).</p>
<p>However, you can do even better: consider the tree of height <code>5</code> in the middle of the fourth row:</p>
<pre><code>30373
25512
65332
33<em>5</em>49
35390
</code></pre>
<ul>
<li>Looking up, its view is blocked at <code><em>2</em></code> trees (by another tree with a height of <code>5</code>).</li>
<li>Looking left, its view is not blocked; it can see <code><em>2</em></code> trees.</li>
<li>Looking down, its view is also not blocked; it can see <code><em>1</em></code> tree.</li>
<li>Looking right, its view is blocked at <code><em>2</em></code> trees (by a massive tree of height <code>9</code>).</li>
</ul>
<p>This tree's scenic score is <code><em>8</em></code> (<code>2 * 2 * 1 * 2</code>); this is the ideal spot for the tree house.</p>
<p>Consider each tree on your map. <em>What is the highest scenic score possible for any tree?</em></p>
</article>


In [ ]:
from more_itertools import takewhile_inclusive


class Grid:
    def __init__(self, s: str) -> None:
        self.grid = [[int(c) for c in l.strip()] for l in s.strip().splitlines()]
        self.rows, self.cols = len(self.grid), len(self.grid[0])

    def count(self, row: int, col: int, rows: range, cols: range) -> int:
        check = lambda c: self.grid[c[0]][c[1]] < self.grid[row][col]
        return sum(1 for _ in takewhile_inclusive(check, product(rows, cols)))

    def scenic_score(self, row: int, col: int) -> int:
        res = self.count(row, col, range(row, row + 1), range(col - 1, -1, -1))
        res *= self.count(row, col, range(row, row + 1), range(col + 1, self.cols))
        res *= self.count(row, col, range(row - 1, -1, -1), range(col, col + 1))
        res *= self.count(row, col, range(row + 1, self.rows), range(col, col + 1))
        return res

    def highest_scenic_score(self) -> int:
        return max(
            self.scenic_score(r, c)
            for r, c in product(range(self.rows), range(self.cols))
        )

    def __str__(self) -> str:
        return "\n".join(f"{''.join(str(i) for i in t)}" for t in self.grid)


tests = [
    {
        "name": "Example",
        "s": """
            30373
            25512
            65332
            33549
            35390
        """,
        "expected": 8,
    },
]


@test(tests=tests)
def test_part_II(s: str) -> int:
    return Grid(s).highest_scenic_score()


Test Example passed, for test_part_II.
Success


<link href="style.css" rel="stylesheet"></link>


In [53]:
print(f"Part 2: {Grid(puzzle).highest_scenic_score()}")

Part 2: 201600


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>201600</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
